wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -90,80 +114,118 @@ impl LastVotedForkSlotsAggregate {
-                    return None;
+                    return true;
                 } else {
-                    HashSet::from_iter(old_slots.to_slots(self.root_slot))
+                    HashSet::from_iter(old_slots.to_slots(self.root_bank.slot()))
                 }
             }
             None => HashSet::new(),
         };
         for slot in old_slots_set.difference(&new_slots_set) {
+            let epoch = self.root_bank.epoch_schedule().get_epoch(*slot);
             let entry = self.slots_stake_map.get_mut(slot).unwrap();
-            *entry = entry.saturating_sub(sender_stake);
-            if *entry < threshold_stake {
-                self.slots_to_repair.remove(slot);
+            if let Some(sender_stake) = self.root_bank.epoch_node_id_to_stake(epoch, from) {
+                *entry = entry.saturating_sub(sender_stake);
+                let repair_threshold_stake = (self.root_bank.epoch_total_stake(epoch).unwrap()
+                    as f64
+                    * self.repair_threshold) as u64;
+                if *entry < repair_threshold_stake {
+                    self.slots_to_repair.remove(slot);
+                }
             }
         }
         for slot in new_slots_set.difference(&old_slots_set) {
+            let epoch = self.root_bank.epoch_schedule().get_epoch(*slot);
             let entry = self.slots_stake_map.entry(*slot).or_insert(0);
-            *entry = entry.saturating_add(sender_stake);
-            if *entry >= threshold_stake {
-                self.slots_to_repair.insert(*slot);
+            if let Some(sender_stake) = self.root_bank.epoch_node_id_to_stake(epoch, from) {
+                *entry = entry.saturating_add(sender_stake);
+                let repair_threshold_stake = (self.root_bank.epoch_total_stake(epoch).unwrap()
+                    as f64
+                    * self.repair_threshold) as u64;
+                if *entry >= repair_threshold_stake {
+                    self.slots_to_repair.insert(*slot);
+                }
             }
         }
-        Some(record)
+        false
+    }
+
+    fn update_epoch_info_active_stake(&mut self) {
+        for entry in self.epoch_info_vec.iter_mut() {
+            entry.total_active_stake = self.active_peers.iter().fold(0, |sum: u64, pubkey| {
+                sum.saturating_add(
+                    self.root_bank
+                        .epoch_node_id_to_stake(entry.epoch, pubkey)
+                        .unwrap_or(0),
+                )
+            });
+        }
     }
 
-    pub(crate) fn active_percent(&self) -> f64 {
-        let total_stake = self.epoch_stakes.total_stake();
-        let total_active_stake = self.active_peers.iter().fold(0, |sum: u64, pubkey| {
-            sum.saturating_add(Self::validator_stake(&self.epoch_stakes, pubkey))
-        });
-        total_active_stake as f64 / total_stake as f64 * 100.0
+    fn update_epoch_info_considered_during_exit(&mut self) {
+        let highest_repair_slot_epoch = self
```

Comment:
Right now one malicious peer voting with a much higher slot could cause the entire cluster to add the next epoch's stakes to consider.

Is there any benefit to only considering the next epoch's stakes if:

Let's say we have 100 stake in first epoch

1. In the next epoch, the previous epoch's valdiators should have at least 91 stake (assuming worst case 9% new stake and 9% old stake cooldown).
3. If next epoch has OC'd, 67 of next epoch's eshould have voted, of which at least 67 + 91 - 100 = 58 stake is from the previoius epoch.

Thus we could say we ignore the next epoch until at least 58% of the previous epoch's stake has voted for the next epoch. Not sure if this provides more security.

Reply:
One malicious peer voting shouldn't be a problem, we only consider if any slot has >42% of the stake, that should be a lot of validators already.

This 42% limit is just very convenient because we use it for repair as well. We could change it to 58% or 38% with different tradeoffs in mind.  I think your proposal provide similar safety guarantee to the current requirement (>=1 slot with >42% epoch stake), but when you say 58% you didn't count the stake outside the restart. So if we have 80% of the old epoch stake joining the restart, then we have at least (80-9)/(100+9) = 65% of the new epoch stake, 65% - 20% = 45% of the old Epoch stake could lead to us waiting for new Epoch. We could change the
`> 1 slot on 42%` limit to `all stake >45%`, but I would think it doesn't make too much difference in reality.

Reply:
Hmm, now I think the limit should be at least 1 slot with >=33% stake in the new Epoch.

Let's say some slot X is OC'ed in the new epoch, it has >= 67% stake in the new epoch, we can have at most 9% churn, so it should have >=58% stake in the old epoch. Let's say we have >=80% of the old Epoch in the restart, and 5% can lie, then any slot OC'ed in the new epoch should have at least 58-(100-80)-5 = 33% stake when the old Epoch has >=80% stake in restart.

That said, if we are this close to the new epoch we might very soon need 80% of the new Epoch stake to be online for the cluster to continue functioning.

I'm now thinking, in the spirit of guaranteeing if the clock has ticked into the new Epoch for enough people then we require 80% stake to be online for the new epoch, how about waiting for the new Epoch as soon as you have 1/3 of the stake voting into the new cluster? We would never have >=1/3 malicious validator, right?

Reply:
Changed to check that >1/3 voted for at least 1 slot in the new Epoch.

If enough people have their clocks close to the new Epoch boundary, we probably should wait for >80% stake so we can reach consensus in the new Epoch.